File: app/src/main/java/starbright/com/projectegg/dagger/module/AppModule.java
package starbright.com.projectegg.dagger.module;

import android.app.Application;
import android.arch.persistence.room.Room;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.IOException;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import id.zelory.compressor.Compressor;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import starbright.com.projectegg.BuildConfig;
import starbright.com.projectegg.dagger.qualifier.ApplicationContext;
import starbright.com.projectegg.data.local.RecipeDatabase;
import starbright.com.projectegg.util.scheduler.BaseSchedulerProvider;
import starbright.com.projectegg.util.scheduler.SchedulerProvider;

/**
 * Created by Andreas on 4/8/2018.
 */

@Module
public class AppModule {
    private Application mApplication;

    public AppModule(Application application) {
        mApplication = application;
    }

    @Provides
    @ApplicationContext
    @Singleton
    Context provideContext() { return mApplication; }

    @Provides
    @Singleton
    BaseSchedulerProvider providesSchedulerComposer() {
        return SchedulerProvider.getInstance();
    }

    @Provides
    @Singleton
    Compressor providesCompressor(@ApplicationContext Context context) {
        return new Compressor(context);
    }
}

File: app/src/main/java/starbright/com/projectegg/dagger/module/AppModule.kt
/*
 * Copyright (c) by Andreas (oentoro.andreas@gmail.com)
 * created at 25 - 7 - 2020.
 */

/**
 * Created by Andreas on 4/8/2018.
 */

package starbright.com.projectegg.dagger.module

import android.app.Application
import android.content.Context
import dagger.Module
import dagger.Provides
import id.zelory.compressor.Compressor
import io.reactivex.disposables.CompositeDisposable
import starbright.com.projectegg.dagger.qualifier.ApplicationContext
import starbright.com.projectegg.util.NetworkHelper
import starbright.com.projectegg.util.scheduler.RxSchedulerProvider
import starbright.com.projectegg.util.scheduler.SchedulerProviderContract
import javax.inject.Singleton

@Module
class AppModule(private val application: Application) {

    @Provides
    @ApplicationContext
    fun provideContext(): Context {
        return application
    }

    @Provides
    @Singleton
    fun providesNetworkHelper(): NetworkHelper = NetworkHelper(application)

    @Provides
    @Singleton
    fun providesCompressor(@ApplicationContext context: Context): Compressor {
        return Compressor(context)
    }

    @Provides
    fun providesCompositeDiposable(): CompositeDisposable {
        return CompositeDisposable()
    }

    @Provides
    @Singleton
    fun providesSchedulerComposer(): SchedulerProviderContract {
        return RxSchedulerProvider()
    }
}
